File: totvs-camera/src/main/java/com/totvs/camera/impl/ImageProxyImpl.kt
package com.totvs.camera.impl

import android.graphics.ImageFormat
import android.graphics.Rect
import android.media.Image
import androidx.annotation.RestrictTo
import com.totvs.camera.core.ImageInfo
import com.totvs.camera.core.ImageProxy
import java.nio.ByteBuffer

/**
 * Concrete implementation of [ImageProxy]
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal class ImageProxyImpl(
    override val image: Image?,
    private  val rotationDegrees: Int
) : ImageProxy {

    override var cropRect: Rect
        get() = image?.cropRect ?: Rect(0, 0, 0, 0)
        set(value) {
            image?.cropRect = value
        }

    override val format: Int
        get() = image?.format ?: ImageFormat.UNKNOWN

    override val width: Int
        get() = image?.width ?: 0

    override val height: Int
        get() = image?.height ?: 0

    override val imageInfo: ImageInfo
        get() = ImageInfoImpl(
            timestamp = image?.timestamp ?: 0,
            rotationDegrees = rotationDegrees
        )

    override val planes: List<ImageProxy.PlaneProxy>
        get() = (image?.planes ?: emptyArray()).map {
            PlaneProxyImpl(buffer = it.buffer)
        }

    override fun close() {
        image?.close()
    }

    private class PlaneProxyImpl(override val buffer: ByteBuffer) : ImageProxy.PlaneProxy
}

File: ios/TOTVSCameraKit/View/Core/ImageProxyImpl.swift
//
//  ImageProxyImpl.swift
//  TOTVSCameraKit
//
//  Created by Jansel Rodriguez on 7/9/20.
//  Copyright © 2020 Jansel. All rights reserved.
//

import UIKit
import CoreMedia

/**
* Concrete implementation of [ImageProxy]
*/
class ImageProxyImpl : ImageProxy {
    
    let imageInfo: ImageInfo
    
    var image: CIImage?
    
    var buffer: CMSampleBuffer?
    
    var width: Int
    
    var height: Int
    
    init(image: CIImage, width: Int, height: Int, imageInfo: ImageInfo) {
        self.image = image
        self.buffer = nil
        self.width = width
        self.height = height
        self.imageInfo = imageInfo
    }
    
    init(buffer: CMSampleBuffer, width: Int, height: Int, imageInfo: ImageInfo) {
        self.image = nil
        self.buffer = buffer
        self.width = width
        self.height = height
        self.imageInfo = imageInfo
    }
    
    
    func close() {
        image = nil
        buffer = nil
    }
    
    deinit {
        close()
    }
}
